src/HttpRequest.cc

```diff
@@ -738,39 +738,59 @@ HttpRequest::manager(const CbcPointer<ConnStateData> &aMgr, const AccessLogEntry
     }
 }
 
-/// cacheIp and localIp are useful when required information may not be available in request
-/// \see const Ip::Address *FindListeningPortAddress(const HttpRequest *request, const AccessLogEntry *al)
+/// a helper for validating FindListeningPortAddress()-found address candidates
 static const Ip::Address *
-FindListeningPortAddress(const HttpRequest *request, const Ip::Address *cacheIp, const Ip::Address *localIp)
-{
-    if (request) {
-        if (request->flags.interceptTproxy || request->flags.intercepted) {
-            if (request->masterXaction->squidPort) {
-                if (!request->masterXaction->squidPort->s.isAnyAddr())
-                    return &(request->masterXaction->squidPort->s);
-                else if (!cacheIp)
-                    return nullptr;
-            }
-            if (cacheIp)
-                return (cacheIp->isAnyAddr() ? nullptr : cacheIp);
-	    // XXX: is it correct to return localIp if cacheIp didn't have info?
-        }
-        // XXX: do we need to check isAnyAddr() here?
-        if (request->masterXaction->tcpClient)
-            localIp = &(request->masterXaction->tcpClient->local);
-    }
-    // XXX: do we need to check isAnyAddr() here?
-    return localIp;
+FindListeningPortAddressInAddress(const Ip::Address *ip)
+{
+    // FindListeningPortAddress() callers do not want "any" addresses
+    return (ip && !ip->isAnyAddr()) ? ip : nullptr;
+}
+
+/// a helper for handling PortCfg cases of FindListeningPortAddress()
+static const Ip::Address *
+FindListeningPortAddressInPort(const AnyP::PortCfgPointer &port)
+{
+    return port ? FindListeningPortAddressInAddress(&port->s) : nullptr;
+}
+
+/// a helper for handling Connection cases of FindListeningPortAddress()
+static const Ip::Address *
+FindListeningPortAddressInConn(const Comm::ConnectionPointer &conn)
+{
+    return conn ? FindListeningPortAddressInAddress(&conn->local) : nullptr;
 }
 
 const Ip::Address *
-FindListeningPortAddress(const HttpRequest *request, const AccessLogEntry *al)
-{
-    if (al)
-        return FindListeningPortAddress(
-                    request ? request : al->request, // use request if specifically given
-                    al->cache.port ? &(al->cache.port->s) : nullptr,
-                    al->tcpClient ? &(al->tcpClient->local) : nullptr);
-    return FindListeningPortAddress(request, nullptr, nullptr);
+FindListeningPortAddress(const HttpRequest *callerRequest, const AccessLogEntry *ale)
+{
+    // Check all sources of usable listening port information, giving
+    // HttpRequest and masterXaction a preference over ALE.
+
+    const HttpRequest *request = callerRequest;
+    if (!request && ale)
+        request = ale->request;
+    if (!request)
+        return nullptr; // not enough information
+
+    const Ip::Address *ip = nullptr;
+
+    if (request->flags.interceptTproxy || request->flags.intercepted) {
+        if (!ip)
```

Comment:
We dont need to check ip here - it will indeed be nullptr.

Reply:
I know, but I favor more readable code more than eliminating on extra pointer check. Right now, the structure of this code is very simple/clear. If you remove this `if`, it becomes a little bit trickier to follow the overall address searching pattern.

I hope the compiler will remove this check rather than complain that it is "pointless".

Please feel free to remove it if you do not like it.

Reply:
Actually I am hesitant of too many ``if``s in code. For my eyes that makes code reading slow (but I understand its personal choice).

I will make slight changes to flow in a while. Which will reduce 1 ``if`` but code flow would remain similar. (top part and bottom part). Please review once done. Thanks.

Reply:
> I am hesitant of too many ifs in code.

Too many is never good. If you can improve this code, please do!